### PR TITLE
New API xrt::hw_context

### PR DESCRIPTION
**Pull request by stsoe**
New concept to represent hardware resources (columns, regions, PS
processes) used by an xclbin.  A hardware context is used by an
application to control where execution takes place.  In a multi-xclbin
scenario, the hardware context is specific to one xclbin even if that
xclbin is loaded multiple times.

The feature is optional for legacy applications / platforms.

Constructed from a device and an xclbin uuid that has been registered
with the device (driver)

A context is 1-1 with a slot index. A resource resolver assigns a
hardware context to hardware resources, maybe sharing, or fails.
On success, the driver assigns a unique slot index to the context.

A kernel object is created from a hardware context. CU context
creation no longer needs EAGAIN as opening of CU context always
succeeds if the hardware context was successfully created. The rather
complicated code for EAGAIN per #6421 has been removed in this PR.

A hardware context can be shared between threads. A process
can construct multiple hardware contexts.

### CODE REVIEW

File: tests/xrt/multi_xclbin/simple.cpp [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+
+#include "experimental/xrt_hw_context.h"
+
+#include <array>
+#include <atomic>
+#include <cstdlib>
+#include <fstream>
+#include <iostream>
+#include <list>
+#include <thread>
```
Comment (keryell) [inline tests/xrt/multi_xclbin/simple.cpp:21]: Eek! Be modest in your inclusions. :-) Or is this from the `#include <C++>` diversity&inclusion group?

File: tests/xrt/multi_xclbin/simple.cpp [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+
+#ifdef _WIN32
+# pragma warning ( disable : 4267 )
+#endif
+
+// Kernel specifics
+// void addone (__global ulong8 *a, __global ulong8 * b, unsigned int  elements)
+// addone(a, b, ELEMENTS)
+// The kernel is compiled with 8 CUs same connectivity.
+static constexpr size_t elements = 16;
```
Comment (keryell) [inline tests/xrt/multi_xclbin/simple.cpp:32]: Do you really need all these `static` everywhere in a such single-file example? Same for the functions.
Nowadays the anonymous `namespace` might be more fashionable for this anyway.

File: tests/xrt/multi_xclbin/simple.cpp [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+static constexpr size_t array_size = 8;
+static constexpr size_t maxcus = 8;
```
Comment (keryell) [inline tests/xrt/multi_xclbin/simple.cpp:34]:
```suggestion
constexpr size_t max_cus = 8;
```

**Review by keryell — CHANGES_REQUESTED** [review]
Quite interesting feature!

**Comment by keryell** [comment]
@Ralender context! How does this impact the future of https://github.com/triSYCL/sycl/pull/170 ?

**Comment by stsoe** [comment]
@saifuddin-xilinx @chvamshi-xilinx   I added you both as reviewer, most FYI.  I don't think the changes in this PR affects vck5000 or builtin xclbin right now.   We will meet some time this week to summarize the new APIs.  The slides (spec) are a bit like work-in-progress, but you should get the the idea by checking the slide about the new xcl APIs that are being added.

**Review by saifuddin-xilinx — APPROVED** [review]

**Pull request merged**